### PR TITLE
add 'include keyboard_features.mk' into build_keyboard.mk

### PR DESCRIPTION
**Pull request by mtei**
This is an alternative implementation of #6794 and #2046.

## Description

`keyboard_features.mk` is a file containing keyboard post-processing rules.

Post-processing rules convert keyboard-specific shortcuts that represent combinations of standard options into QMK standard options.

`keyboard_features.mk` is a keyboard-local version of the functions performed by `common_features.mk`.

`rules.mk` and `keyboard_features.mk` are processed in the following order:
```
keyboards/KEYBOARD_NAME/rules.mk
keyboards/KEYBOARD_NAME/rev1/rules.mk

keyboards/KEYBOARD_NAME/keymap/KEYMAP_NAME/rules.mk
users/USER_NAME/rules.mk

keyboards/KEYBOARD_NAME/rev1/keyboard_features.mk. (this PR ADD)
keyboards/KEYBOARD_NAME/keyboard_features.mk       (this PR ADD)
common_features.mk
```

### Example 1
The `keyboard_features.mk` file can interpret `features` of a keyboard-level before `common_features.mk`.  For example, when your designed keyboard has the option to implement backlighting or underglow using rgblight.c, writing the following in the `keyboard_features.mk` makes it easier for the user to configure the `rules.mk`.

* `keyboards/top_folder/keymaps/a_keymap/rules.mk`
  ```makefile
  # Please set the following according to the selection of the hardware implementation option.
  RGBLED_OPTION_TYPE = backlight   ## none, backlight or underglow
  ```
* `keyboards/top_folder/keyboard_features.mk`
  ```makefile
  ifeq ($(filter $(strip $(RGBLED_OPTION_TYPE))x, nonex backlightx underglowx x),)
     $(error unknown RGBLED_OPTION_TYPE value "$(RGBLED_OPTION_TYPE)")
  endif

  ifeq ($(strip $(RGBLED_OPTION_TYPE)),backlight)
    RGBLIGHT_ENABLE = yes
    OPT_DEFS += -DRGBLED_NUM=30
  endif
  ifeq ($(strip $(RGBLED_OPTION_TYPE)),underglow)
    RGBLIGHT_ENABLE = yes
    OPT_DEFS += -DRGBLED_NUM=6
  endif
  ```

### Example 2
Now, if you compile the following, you get an error.

```shell
make plaid:bakingpy
.......
light/backlight_driver_common.c:6:6: error: #error "Backlight pin/pins not defined. Please configure."
 #    error "Backlight pin/pins not defined. Please configure."
      ^~~~~
quantum/backlight/backlight_driver_common.c:22:36: error: 'BACKLIGHT_PIN' undeclared here (not in a function); did you mean 'BACKLIGHT_ENABLE'?
 static const pin_t backlight_pin = BACKLIGHT_PIN;
                                    ^~~~~~~~~~~~~
                                    BACKLIGHT_ENABLE    [ERRORS]
 | 
 | 
 | 
make[1]: *** [.build/obj_plaid_bakingpy/quantum/backlight/backlight_driver_common.o] Error 1
make: *** [plaid:bakingpy] Error 1
Make finished with errors
```

The reason for the error is that `layouts/community/ortho_4x12/bakingpy/rules.mk` does not have a `plaid` keyboard check.

```makefile
ifneq ($(LAYOUTS_HAS_RGB), no)
  RGBLIGHT_ENABLE = yes
endif
AUDIO_ENABLE = no
ifeq ($(strip $(KEYBOARD)), zlant)
  BACKLIGHT_ENABLE = no
else ifeq ($(strip $(KEYBOARD)), 40percentclub/4x4)
  BACKLIGHT_ENABLE = no
else ifneq (, $(findstring lets_split, $(KEYBOARD)))
  BACKLIGHT_ENABLE = no
else
  BACKLIGHT_ENABLE = yes
endif
```

If you write the following in `keyboards/plaid/keyboard_features.mk`, the above compilation error will be resolved.

```makefile
BACKLIGHT_ENABLE = no
```

You should write the same thing as above in the `keyboard_features.mk` for each keyboard of `zlant`, `40percentclub/4x4`, and `lets_split`. Then `layouts/community/ortho_4x12/bakingpy/rules.mk` can be as simple as this:

```makefile
ifneq ($(LAYOUTS_HAS_RGB), no)
  RGBLIGHT_ENABLE = yes
endif
AUDIO_ENABLE = no
BACKLIGHT_ENABLE = yes
```

## Note
Drashna said about #6794:

> I think that seeb's may be the better implementation, and the one that we should go with, long term. It supports every layer of the makefile stuff, which also allows for some pretty cool stuff.

Yes, #6794 is a bit rough implementation. However, #2046 was an implementation that was too difficult for me to understand.

So, in this PR, I tried again a straightforward and general implementation.

The Helix keyboard really need this PR feature. That's because it turned out yesterday that the QMK Configurator did not compile Helix firmware correctly since October 2019(#6952). (https://github.com/qmk/qmk_configurator/issues/684) (https://github.com/qmk/qmk_configurator/issues/714)

On branch https://github.com/mtei/qmk_firmware/tree/update_helix_rules_mk_using_keyboard_features_mk, I used this feature to modify rules.mk for all Helix keymaps, and confirmed that the compilation did not change.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/qmk/qmk_configurator/issues/684 (In addition, https://github.com/mtei/qmk_firmware/tree/update_helix_rules_mk_using_keyboard_features_mk  is required)

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by drashna** [comment]
One thing I would like to add here. If we're moving stuff around, I think this needs a userspace level needs a second pass check too.  And for much the same reasons that the helix style boards too, in fact. 

And with the userspace check changed, so that it is consistent with specific to generic (keymap <-> user <-> keyboard). Specifically, to be more in line with the config.h handling.  

Eg: 
```
keyboards/KEYBOARD_NAME/rules.mk
keyboards/KEYBOARD_NAME/rev1/rules.mk
users/USER_NAME/rules.mk
keyboards/KEYBOARD_NAME/keymap/KEYMAP_NAME/rules.mk
keyboards/KEYBOARD_NAME/keymap/KEYMAP_NAME/keyboard_features.mk
keyboards/KEYBOARD_NAME/rev1/keyboard_features.mk
_users/USER_NAME/keyboard_features.mk_
keyboards/KEYBOARD_NAME/keyboard_features.mk
common_features.mk
```


Aside from this, I very much do like the change

**Comment by mtei** [comment]
@drashna, I understand the significance of your hope.

I looked it up. 

```shell
% find users keyboards/keebio -name keyboard_features.mk
users/drashna/keyboard_features.mk
keyboards/keebio/iris/keymaps/drashna/keyboard_features.mk
keyboards/keebio/iris/keymaps/keyboard_features.mk
keyboards/keebio/iris/rev4/keyboard_features.mk
keyboards/keebio/iris/keyboard_features.mk
keyboards/keebio/keyboard_features.mk

% make keebio/iris/rev4:drashna:clean
QMK Firmware 0.8.37
==== this is keyboards/keebio/iris/rev4/rules.mk
==== this is keyboards/keebio/rules.mk
==== this is keyboards/keebio/iris/rules.mk
==== this is keyboards/keebio/iris/rev4/rules.mk
Making keebio/iris/rev4 with keymap drashna and target clean

== start build_keyboard.mk ====
==== this is keyboards/keebio/rules.mk
==== this is keyboards/keebio/iris/rules.mk
==== this is keyboards/keebio/iris/rev4/rules.mk
==== this is keyboards/keebio/iris/keymaps/drashna/rules.mk
==== this is users/drashna/rules.mk
==== this is keyboards/keebio/iris/rev4/keyboard_features.mk
==== this is keyboards/keebio/iris/keyboard_features.mk
==== this is keyboards/keebio/keyboard_features.mk
                 (((((Easy to insert here)))))
==== this is common_features.mk
== end build_keyboard.mk ====
== include tmk_core/rules.mk
```

It seems difficult for build_keyboard.mk to decide where to insert `users/USER_NAME/keyboard_features.mk`.

see https://github.com/mtei/qmk_firmware/tree/test_user_rules

**Comment by mtei** [comment]
@drashna

I was a bit misunderstood how build_keyboard.mk works.

In this PR, it will be processed as follows.
```
keyboards/KEYBOARD_NAME/rules.mk
keyboards/KEYBOARD_NAME/rev1/rules.mk
                                                   (drashna need HERE?)
keyboards/KEYBOARD_NAME/keymap/KEYMAP_NAME/rules.mk
users/USER_NAME/rules.mk

keyboards/KEYBOARD_NAME/rev1/keyboard_features.mk. (this PR ADD)
keyboards/KEYBOARD_NAME/keyboard_features.mk       (this PR ADD)
common_features.mk
```

**Comment by mtei** [comment]
@drashna , What do you think if it looks like this?

```
keyboards/KEYBOARD_NAME/rules.mk
keyboards/KEYBOARD_NAME/rev1/rules.mk

users/USER_NAME/user_rules.mk                      (NEW)
keyboards/KEYBOARD_NAME/keymap/KEYMAP_NAME/rules.mk
users/USER_NAME/rules.mk                           (Leave for backward compatibility)

keyboards/KEYBOARD_NAME/rev1/keyboard_features.mk. (this PR ADD)
keyboards/KEYBOARD_NAME/keyboard_features.mk       (this PR ADD)
users/USER_NAME/user_features.mk                   (NEW)

common_features.mk
```

**Comment by drashna** [comment]
I'm not a fan of leaving the `rules.mk` and adding a `user_rules.mk`, as it breaks from the normal naming convention.  

If anything, I'd say just rename them, and have everything setup "correctly", but that would mean going through a breaking change. But that may be the best way to go about this.

**Comment by mtei** [comment]
If the change for 'users/USER_NAME/*. mk' is a breaking change, I'd like to make that change in another PR.

I now need this PR change to resolve qmk/qmk_configurator#684.

**Comment by vomindoraan** [comment]
I agree with @drashna. Userspace `rules.mk` being processed after keymap `rules.mk` has been a major pet peeve of mine for a long time. I believe we will need a breaking change to be able to address that, though.

**Comment by mtei** [comment]
@vomindoraan and @drashna, I agree.

However, I would like to focus on the addition of `keyboard_features.mk` in this PR.

I think the improvement of `rules.mk` for user-level and the addition of `user_features.mk` should be proposed in a separate PR.

**Comment by mtei** [comment]
@skullydazed @drashna, do you have any objections/opinions to this PR?

**Comment by mtei** [comment]
> For example 2, I think over time we have conflated the idea of enabling a feature with the idea that a feature is supported. Features should degrade gracefully if a keymap uses a feature a keyboard can't support. I have an alternative proposal to support this use case:

I agree with your suggestion in this case. It seems to be better.

**Comment by mtei** [comment]
> For example 1, that seems awfully complicated. I don't have a solution for this use case yet but I am still considering the problem.

Ever since #2606, the Helix keyboard has been in need of this feature.

I've been waiting for #2046 or #6794 or this #8422.

And, with (qmk/qmk_configurator#684) and (qmk/qmk_configurator#714) reporting that Helix firmware cannot be generated, this feature is more strongly needed.

If I significantly rewrite the rules.mk and config.h under keyboards/helix/rev2/keymaps, we might be able to generate Helix firmware, even under the qmk_configurator environment. However, it is a breaking chang, so I want to avoid it.

**Comment by skullydazed** [comment]
I've been thinking a lot about this PR. I'm worried that it increases complexity given the future direction[1] of QMK. I'd like to start a conversation to determine the best path for helix going forward.

We have been working toward a data driven infrastructure so that tools like QMK Configurator can give users an easy way to build a firmware. Work I'm doing in the develop branch right now lays the foundation for API being able to control many features inside QMK. Part of this is moving configuration out of `rules.mk` and `config.h` and into `info.json`. My hope is this will make it easier for non-programmers to maintain keyboards, and that someone will write tools to make that even easier. 

I love what you're doing to make it easy to compile helix, but unfortunately I don't see a way to align this PR with the data driven work. The good news is that as part of the data driven work there will be a mechanism to combine snippets of JSON, allowing you to create a similar system for helix. For example, a user could specify in their `keymap.json` file that they want `oled.json`, `back.json`, and `ios.json` included to pull in those features.

I'm not sure what the best short term solution is. I think it's fine to let things continue as they currently are if you're happy with the status quo. The `HELIX=` system won't work with configurator, but as a workaround you can create a sub-folder that sets some default options. I think this may be what the `qmk_conf` subfolder is?

1. https://hackmd.io/fpPylbKgQROie6ndKHimfA

**Comment by tzarc** [comment]
@skullydazed can we get some clarification on direction, here? This seems to be somewhat complete -- should we hold up stuff like this for a future hypothetical implementation that may never arrive?

**Comment by tzarc** [comment]
So, given that the intent was to describe use-cases, I'll outline a specific example.

This is the current state of affairs on `master`:
image

For the Djinn, which has optional USBPD support available, I'd intended to add a variable to `rules.mk` as per `common_features.mk`-style subsystem enabling -- `USBPD_ENABLE` -- with the standard `yes`/`no` choice.

The intent was that this would then be used to set up some preprocessor defines through `OPT_DEFS` and the like, based on the user selection.

As it stands, without functionality that this provides, the only way for those `OPT_DEFS` defines to be set after the keymap's `rules.mk` would be to include it in the global  `common_features.mk`... and this was why we got #11824. The PR checklist suggests that code that is candidate for core should be prototyped and proven within keyboards themselves, but that's not possible when requiring modifications through `rules.mk` due to ordering.

This clearly isn't the first instance of a keyboard preferring to set things "internally" based on the selections of a keymap, but unfortunately there's no sequence of `rules.mk` equivalents in order to achieve this, without merging it into core's `common_features.mk`.

**Comment by skullydazed** [comment]
I spent some time talking with @tzarc about this.

(Quick aside- when I say "processing rules" below I'm talking about any makefile content (`rules.mk`) that isn't a simple variable assignment.)

During the conversation we determined that my primary concern is around the current state of the repo- we already have a lot of `rules.mk` files with processing rules in them, and trying to explain why you put those processing rules in `rules.mk` or `keyboard_features.mk` is going to go over most keyboard designer's heads. I also fear that most of them are just going to use `rules.mk` anyway and not allow for overriding at the keymap level at all.

To solve this I think we should explicitly move all processing rules to a separate file. We could call this file `processing.mk` (or another descriptive name) to make its purpose clear. This file will live at the keyboard level and always be processed after all `rules.mk` files. This will give us a clean separation between configuration (which can be set at either the keyboard or keymap level) and logic, and will ensure that all makefile logic sees the same configuration.

We can extend `qmk lint` to enforce this separation, so that if a keyboard maintainer does put logic into their `rules.mk` it will be caught at PR time.

As a path forward we can retarget this PR at `develop`, modify it to use the name we decide on (`processing.mk` or something else,) extend the `qmk lint` command, and then get both merged. I assume that I will extend `qmk lint` while mtei will modify this PR, but that can be changed.

@mtei what do you think about this plan?

**Comment by tzarc** [comment]
I'll also point out that if we substitute your `processing.mk` with `build.mk` we basically get #2046, too.

**Comment by mtei** [comment]
> I'd like to start a conversation to determine the best path for helix going forward.

Sorry for the delay in replying.

> The good news is that as part of the data driven work there will be a mechanism to combine snippets of JSON, allowing you to create a similar system for helix. 

I felt like I could count on this, but I didn't have a concrete image of what would happen yet, so I just spent too much time thinking about it to reply.

**What I want to achieve with this PR is Example 1, which I mentioned in the first explanation.**

> The HELIX= system won't work with configurator, but as a workaround you can create a sub-folder that sets some default options. I think this may be what the qmk_conf subfolder is?

The Helix= system is not the subject of this PR, so please do not consider it in this discussion.  The Helix= system is for users like me who run the make command directly from the unix shell.  It was created to temporarily override the keymap-level rules.mk settings to make it easier to try different variations. I think it can be partially replaced by sub-folder.

The qmk_conf subfolder is a temporary workaround for Helix farm build errors on QMK Configurator. In exchange for disabling the keymap level rules.mk, the error is avoided. So please do not consider it in this discussion.

**Comment by mtei** [comment]
> To solve this I think we should explicitly move all processing rules to a separate file. We could call this file processing.mk (or another descriptive name) to make its purpose clear. This file will live at the keyboard level and always be processed after all rules.mk files. This will give us a clean separation between configuration (which can be set at either the keyboard or keymap level) and logic, and will ensure that all makefile logic sees the same configuration.

I think I can accept the proposal. I am in favor of considering it in that direction.

However, there are some concerns.
For example, in `helix/rev2/keymaps/default/rules.mk` there is a section:

```make
ifeq ($(strip $(OLED_ENABLE)), yes)
    SRC += oled_display.c
endif
```

The rule is to compile `keymaps/default/oled_display.c` when `OLED_ENABLE = yes` is set one level above.
Do we move these things from rules.mk as well?

> As a path forward we can retarget this PR at develop, modify it to use the name we decide on (processing.mk or something else,) extend the qmk lint command, and then get both merged. I assume that I will extend qmk lint while mtei will modify this PR, but that can be changed.

I am ready to cooperate.

**Comment by skullydazed** [comment]
> The rule is to compile keymaps/default/oled_display.c when OLED_ENABLE = yes is set one level above.
Do we move these things from rules.mk as well?

My proposal only deals with things that are at the keyboard level. While it would be conceptually cleaner to also support the same files at the keymap level, as a practical matter changing that everywhere this exists already is a lot of work. My preference is to leave the keymap level as it currently is, but if someone wants to take on the conversion work I'd be ok with adopting the same pattern there.

> I am ready to cooperate.

Once we settle on a name you should be clear to proceed. There is weak agreement on `processing.mk`, if that name works for you let's use it. If you have a suggestion besides `processing.mk` and `build.mk` we could consider that as well.

**Comment by mtei** [comment]
I feel that processing.mk is ambiguous.  

Since we have common_features.mk, keyboard_features.mk would be easier to understand. Is this not good enough?

Other candidates

* prepare_build.mk
* post_ processing.mk

**Comment by tzarc** [comment]
we could also consider the following for alignment purposes:
`config.h` + `post_config.h`
`rules.mk` + `post_rules.mk`

**Comment by skullydazed** [comment]
I would prefer to avoid names require prior knowledge to understand. A first-time keyboard maintainer should be able to reason about what the file is for based on the name alone. By that metric I don't like any of the names so far, but I do not have a better suggestion.

Of the proposed names so far `post_rules.mk` comes the closest to the criteria. I worry that keyboard maintainers are going to simply put everything in there, but that can be addressed at PR time for now.

File: docs/hardware_keyboard_guidelines.md
Comment (skullydazed) [inline docs/hardware_keyboard_guidelines.md:151]: The way I read this section says that values in `keyboards/top_folder/post_rules.mk` override values in `keyboards/top_folder/sub_1/post_rules.mk`. Is that how the code actually behaves?

Comment (mtei) [reply docs/hardware_keyboard_guidelines.md:151]: I envision the following usage.

For example, let's assume that top_folder supports the following two switch macros
* FEATURES_1_ENABLE
* FEATURES_2_ENABLE

In top_folder/post_rules.mk, the following will be handled.
```makefile
ifeq ($(strip $(FEATURE_1_ENABLE)), yes)
   # Configuration for common_features.mk
endif
ifeq ($(strip $(FEATURE_2_ENABLE)), yes)
   # Configuration for common_features.mk
endif
```

Now, let's say we add top_folder/sub1, where we want to add a shortcut FEATURE_ALL_ENABLE.

The file top_folder/sub1/post_rules.mk should look like this
```makefile
ifeq ($(strip $(FEATURE_ALL_ENABLE)), yes)
   FEATURE_1_ENABLE = yes
   FEATURE_2_ENABLE = yes
endif
```

Comment (skullydazed) [reply docs/hardware_keyboard_guidelines.md:151]: Sorry for the delayed response here, I realized I didn't answer while doing a tab sweep.

That's not what I'm asking about here. I believe this list does not match the actual processing done here: <https://github.com/qmk/qmk_firmware/pull/8422/files/e396557341533bb77dc0b0a2b2fa21a774d898e1#diff-94923ddf56c801cf592868f3d0560d912344285dbc067dec0cfb0e2ec0034f28R325-R340>

This list says that `keyboards/top_folder/sub_1/post_rules.mk` is executed first and `keyboards/top_folder/post_rules.mk` is executed second, but the code has that order reversed.

Comment (mtei) [reply docs/hardware_keyboard_guidelines.md:151]: I'm going to load `rules.mk` and `post_rules.mk` in reverse order. Am I doing something wrong?

https://github.com/mtei/qmk_firmware/blob/e396557341533bb77dc0b0a2b2fa21a774d898e1/build_keyboard.mk#L76-L91

```makefile
# Pull in rules.mk files from all our subfolders
ifneq ("$(wildcard $(KEYBOARD_PATH_5)/rules.mk)","")
    include $(KEYBOARD_PATH_5)/rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_4)/rules.mk)","")
    include $(KEYBOARD_PATH_4)/rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_3)/rules.mk)","")
    include $(KEYBOARD_PATH_3)/rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_2)/rules.mk)","")
    include $(KEYBOARD_PATH_2)/rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_1)/rules.mk)","")
    include $(KEYBOARD_PATH_1)/rules.mk
endif
```

https://github.com/mtei/qmk_firmware/blob/e396557341533bb77dc0b0a2b2fa21a774d898e1/build_keyboard.mk#L325-L340

```makefile
# Pull in post_rules.mk files from all our subfolders
ifneq ("$(wildcard $(KEYBOARD_PATH_1)/post_rules.mk)","")
    include $(KEYBOARD_PATH_1)/post_rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_2)/post_rules.mk)","")
    include $(KEYBOARD_PATH_2)/post_rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_3)/post_rules.mk)","")
    include $(KEYBOARD_PATH_3)/post_rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_4)/post_rules.mk)","")
    include $(KEYBOARD_PATH_4)/post_rules.mk
endif
ifneq ("$(wildcard $(KEYBOARD_PATH_5)/post_rules.mk)","")
    include $(KEYBOARD_PATH_5)/post_rules.mk
endif
```

Comment (tzarc) [reply docs/hardware_keyboard_guidelines.md:151]: @skullydazed do you disagree with reversing the ordering?

Comment (skullydazed) [reply docs/hardware_keyboard_guidelines.md:151]: Yes, I disagree. Unless I'm missing something here it seems that loading them in reverse order will mean that less specific is overriding more specific. It's not clear to me why reversing the ordering would be desirable.

Comment (mtei) [reply docs/hardware_keyboard_guidelines.md:151]: > This list says that keyboards/top_folder/sub_1/post_rules.mk is executed first and keyboards/top_folder/post_rules.mk is executed second, but the code has that order reversed.

There is no discrepancy between the description in `docs/hardware_keyboard_guidelines.md` and the behavior in `build_keyboard.mk`.

```
        * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/post_rules.mk`
      * `keyboards/top_folder/sub_1/sub_2/sub_3/post_rules.mk`
    * `keyboards/top_folder/sub_1/sub_2/post_rules.mk`
  * `keyboards/top_folder/sub_1/post_rules.mk`
* `keyboards/top_folder/post_rules.mk`
* `common_features.mk`
```

```
# Pull in post_rules.mk files from all our subfolders
 ifneq ("$(wildcard $(KEYBOARD_PATH_1)/post_rules.mk)","")
     include $(KEYBOARD_PATH_1)/post_rules.mk
 endif
 ifneq ("$(wildcard $(KEYBOARD_PATH_2)/post_rules.mk)","")
     include $(KEYBOARD_PATH_2)/post_rules.mk
 endif
 ifneq ("$(wildcard $(KEYBOARD_PATH_3)/post_rules.mk)","")
     include $(KEYBOARD_PATH_3)/post_rules.mk
 endif
 ifneq ("$(wildcard $(KEYBOARD_PATH_4)/post_rules.mk)","")
     include $(KEYBOARD_PATH_4)/post_rules.mk
 endif
 ifneq ("$(wildcard $(KEYBOARD_PATH_5)/post_rules.mk)","")
     include $(KEYBOARD_PATH_5)/post_rules.mk
 endif
```

Comment (mtei) [reply docs/hardware_keyboard_guidelines.md:151]: > Yes, I disagree. Unless I'm missing something here it seems that loading them in reverse order will mean that less specific is overriding more specific. It's not clear to me why reversing the ordering would be desirable.

Are you objecting to the following order of execution?

```
        * `keyboards/top_folder/sub_1/sub_2/sub_3/sub_4/post_rules.mk`
      * `keyboards/top_folder/sub_1/sub_2/sub_3/post_rules.mk`
    * `keyboards/top_folder/sub_1/sub_2/post_rules.mk`
  * `keyboards/top_folder/sub_1/post_rules.mk`
* `keyboards/top_folder/post_rules.mk`
* `common_features.mk`
```

Comment (tzarc) [reply docs/hardware_keyboard_guidelines.md:151]: As far as I can tell, yes, that was the dispute.

As a somewhat hypothetical example, three systems `ABC`, `DEF`, and `GHI` all requiring `CANBUS`.

Reversed:
* sub_4: 
  * `if ABC=yes then NEEDS_CANBUS=yes; SRC += abc.c` 
  * `if DEF=yes then NEEDS_CANBUS=yes; SRC += def.c`
  * `if GHI=yes then NEEDS_CANBUS=yes; SRC += ghi.c`
* sub_3:
  * `if NEEDS_CANBUS=yes AND PLATFORM=CHIBIOS then SRC += canbus_chibios.c`
  * `if NEEDS_CANBUS=yes AND PLATFORM=AVR then SRC += canbus_avr.c`

Sequential:
* sub_3:
  * `if (ABC=yes OR DEF=yes OR GHI=yes) AND PLATFORM=CHIBIOS  then SRC += canbus_chibios.c`
  * `if (ABC=yes OR DEF=yes OR GHI=yes) AND PLATFORM=AVR then SRC += canbus_avr.c`
* sub_4:
  * `if ABC=yes then SRC += abc.c`, 
  * `if DEF=yes then SRC += def.c`
  * `if GHI=yes then SRC += ghi.c`

So the question becomes: should all intermediate levels need to check specific overall features, or can required systems be set to yes/no instead and higher levels check for dependencies instead of features?

Comment (skullydazed) [reply docs/hardware_keyboard_guidelines.md:151]: In that specific example shouldn’t NEED_CANBUS be set in rules.mk and then the SRC can be set in post, thus being able to use the same code on both?

(Apologies for the brevity, I’m on mobile)

Comment (skullydazed) [reply docs/hardware_keyboard_guidelines.md:151]: I talked through this a bit with @tzarc on discord, and the conclusion was the current ordering will be fine. I have concerns about the support load if this feature is used a lot, but if it is not the order is arbitrary.

**Comment by skullydazed** [comment]
I talked with @tzarc about our path forward with this PR. It should get merged in parallel with #12108 so that we can catch and properly enforce the split between rules and post_rules. Given the large impact that will have on `qmk lint` (and the full list of lint errors as shown by `qmk generate-api`) it was decided to merge this into develop first thing after the next merge. 

In the interim I will work on splitting the existing `rules.mk` files into `rules.mk` and `post_rules.mk` so that is ready to merge at the same time, and everything will go in clean.

**Comment by mtei** [comment]
>  it was decided to merge this into develop first thing after the next merge.

Is `the next merge`  mean the merge from `develop` to `master` on August 28?

**Comment by skullydazed** [comment]
> Is `the next merge` mean the merge from `develop` to `master` on August 28?

Yes.

**Review by tzarc — APPROVED** [review]

**Pull request merged**
